File: python/hello_world/hello_world.py
# ============================================================================
#  hello_world -- Python (ctypes) port of examples\Vb6\hello_world\hello_world.bas
#  Writes out.pdf with a single centered line of text.
# ============================================================================
import os
import sys
import datetime
import os, sys
# the wrapper lives at <root>/wrappers/python (source checkout) or beside
# the example tree (shipped package) -- find it without a hard-coded path
_d = os.path.dirname(os.path.abspath(__file__))
for _ in range(6):
    for _c in (os.path.join(_d, 'wrappers', 'python'), _d):
        if os.path.isfile(os.path.join(_c, 'lumaspdf.py')):
            sys.path.insert(0, _c)
            break
    else:
        _d = os.path.dirname(_d)
        continue
    break
import lumaspdf as L


@L.TErrorProc
def err_proc(data, errcode, errmsg, errtype):
    if errmsg:
        print(errmsg.decode("latin-1", "replace"))
    return -1  # break processing on error


def main():
    cr = chr(13)
    pdf = L.pdfNewPDF()
    L.pdfSetOnErrorProc(pdf, 0, err_proc)
    if L.pdfCreateNewPDFA(pdf, b"") == 0:  # output file opened later
        L.pdfDeletePDF(pdf)
        return
    L.pdfSetDocInfoA(pdf, L.diCreator, b"Delphi Example project")
    L.pdfSetDocInfoA(pdf, L.diTitle, b"My first PDF output")

    L.pdfAppend(pdf)
    L.pdfSetFontA(pdf, b"Arial", L.fsItalic, 30.0, 1, L.cp1252)
    txt = ("My first PDF output..." + cr + cr + str(datetime.datetime.now()))
    L.pdfWriteFTextA(pdf, L.taCenter, txt.encode("latin-1", "replace"))
    L.pdfEndPage(pdf)

    out_file = os.path.join(os.path.dirname(os.path.abspath(__file__)), "out.pdf")
    if L.pdfHaveOpenDoc(pdf) != 0:
        L.pdfSetOnErrorProc(pdf, 0, L.TErrorProc(0))
        if L.pdfOpenOutputFileA(pdf, out_file.encode("latin-1")) == 0:
            L.pdfDeletePDF(pdf)
            return
        L.pdfSetOnErrorProc(pdf, 0, err_proc)
    if L.pdfCloseFile(pdf) != 0:
        print("OK: " + out_file)

    L.pdfDeletePDF(pdf)


if __name__ == "__main__":
    main()

File: c/hello_world/hello_world.c
/* hello_world -- C port of examples\Vb6\hello_world\hello_world.bas
 * Writes out.pdf with a single centred line of text (PDF/A). */
#include <stdio.h>
#include <string.h>
#include <time.h>
#include "lumaspdf.h"

static SI32 PDF_CALL ErrProc(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType)
{
    (void)Data; (void)ErrCode; (void)ErrType;
    if (ErrMessage) printf("%s\n", ErrMessage);
    return -1; /* break processing on error */
}

/* Directory of the running exe (argv[0]) into out; strip the file name. */
static void exedir(const char* a0, char* out, size_t n)
{
    char* s;
    strncpy(out, a0, n - 1); out[n - 1] = 0;
    s = strrchr(out, '\\'); if (!s) s = strrchr(out, '/');
    if (s) *s = 0; else strcpy(out, ".");
}

int main(int argc, char** argv)
{
    PPDF pdf;
    char dir[1024], outFile[1100], msg[256];
    time_t t = time(NULL);
    (void)argc;

    pdf = pdfNewPDF();
    pdfSetOnErrorProc(pdf, 0, ErrProc);
    if (pdfCreateNewPDFA(pdf, "") == 0) { pdfDeletePDF(pdf); return 1; }

    pdfSetDocInfoA(pdf, diCreator, "Delphi Example project");
    pdfSetDocInfoA(pdf, diTitle, "My first PDF output");

    pdfAppend(pdf);
    pdfSetFontA(pdf, "Arial", fsItalic, 30.0, 1, cp1252);
    _snprintf(msg, sizeof(msg), "My first PDF output...\r\r%s", ctime(&t));
    pdfWriteFTextA(pdf, taCenter, msg);
    pdfEndPage(pdf);

    exedir(argv[0], dir, sizeof(dir));
    _snprintf(outFile, sizeof(outFile), "%s\\out.pdf", dir);

    if (pdfHaveOpenDoc(pdf)) {
        pdfSetOnErrorProc(pdf, 0, NULL);
        if (pdfOpenOutputFileA(pdf, outFile) == 0) { pdfDeletePDF(pdf); return 1; }
        pdfSetOnErrorProc(pdf, 0, ErrProc);
    }
    if (pdfCloseFile(pdf)) printf("OK: %s\n", outFile);

    pdfDeletePDF(pdf);
    return 0;
}
